### PR TITLE
feat: add grpc instrumentation

### LINKED ISSUE #1041: How to handle gRPC repo w.r.t. attribution/licensing
Hello,

I recently created [`opentelemetry-instrumentation-grpc`](https://github.com/hibachrach/opentelemetry-instrumentation-grpc) (Ruby gem forthcoming) which is mostly just a fork of much of the [`opentelemetry-instrumentation-gruf`](https://github.com/open-telemetry/opentelemetry-ruby-contrib/tree/main/instrumentation/gruf) code.

I had a few questions:
- How would folks prefer attribution to be given?
- I didn't preserve the copyright notices at the top of the files as it's not in the open-telemetry org. That said, I'm happy to restore them should that be preferred--I'm not sure what the policy is w.r.t. in-code copyright notices & forking
- Is my choice of MIT license acceptable? Or does it need to maintain the same license as in this repo (Apache 2 IIRC)?

I would have opened a PR to just add the code to this repo but I wasn't sure if I could commit to maintaining it (which the contributing guidelines request).

Thank you!

### PR DESCRIPTION
**Pull request by hibachrach**
See <https://github.com/hibachrach/opentelemetry-instrumentation-grpc/> for previous git history

Closes #1041

### CODE REVIEW

**Review by kaylareopelle — COMMENTED** [review]
Thank you for opening this, @hibachrach! I'm excited to take a deeper look!

As far as I know, this is the first time we've brought a gem into contrib that was originally created outside the repository. There's a few things about consistency in tooling that I'd like to talk about with the SIG as part of the review. More soon.

**Comment by hibachrach** [comment]
Sounds good--happy to align to y'all in any way; no attachment to anything in this PR as it stands.

File: instrumentation/grpc/lib/opentelemetry/instrumentation/grpc.rb
Comment (kaylareopelle) [inline instrumentation/grpc/lib/opentelemetry/instrumentation/grpc.rb:13]:
```suggestion
    # Contains the OpenTelemetry instrumentation for the gRPC gem
```

File: instrumentation/grpc/spec/example/Gemfile
Comment (kaylareopelle) [inline instrumentation/grpc/spec/example/Gemfile:11]: It looks like this example may have been a carryover from the gruf instrumentation! Would you rather update it for the gRPC gem, or remove it and create an issue to add an example later on?

**Review by kaylareopelle — COMMENTED** [review]
Hi @hibachrach! Thanks for your patience! Great to hear you're happy to align with the rest of the repo! 😄 

The main differences we see between your gRPC gem and the rest of the instrumentation are:
* Rspec instead of minitest 
* Standard instead of Rubocop
* Changelog format
* Unique code of conduct 
* RBS

We also want to add this gem to the GitHub actions for this repo to run tests and release the gem.

The SIG had two ideas about how to make the alignment work: 
1. Merge your PR as-is and the approvers/maintainers will submit subsequent PRs to update the tooling to make it compatible with our CI and more closely mirror the other gems. 
2. Make the updates as part of this PR, merging it in once everything is aligned. The approvers/maintainers might make branches off yours to simplify the movement, or we could describe the changes to you and let you take care of them.

I know you've mentioned you have limited time to work on this, so let us know how you'd prefer to move forward. We're really grateful for the work you've done to create the gem and for bringing it into contrib. We want to make things as smooth as we can for you.

p.s. - I left a few suggestions related to some Gruf leftovers I found when looking this over!

File: instrumentation/grpc/README.md
Comment (kaylareopelle) [inline instrumentation/grpc/README.md:58]:
```suggestion
Bug reports and pull requests are welcome on GitHub at https://github.com/open-telemetry/opentelemetry-ruby-contrib. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/CODE_OF_CONDUCT.md).
```

File: instrumentation/grpc/README.md
Comment (kaylareopelle) [inline instrumentation/grpc/README.md:66]:
```suggestion
Everyone interacting in the OpenTelemetry::Instrumentation::Grpc project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/CODE_OF_CONDUCT.md).
```

File: instrumentation/grpc/README.md
Comment (kaylareopelle) [inline instrumentation/grpc/README.md:31]:
```suggestion
  OpenTelemetry::SDK.configure do |c|
    c.use 'OpenTelemetry::Instrumentation::Grpc', {
      peer_service: "Example",
      allowed_metadata_headers: [],
    }
  end
```

File: instrumentation/grpc/README.md
Comment (kaylareopelle) [inline instrumentation/grpc/README.md:40]:
```suggestion
  OpenTelemetry::SDK.configure do |c|
    c.use_all
  end
```

File: instrumentation/grpc/README.md
Comment (kaylareopelle) [inline instrumentation/grpc/README.md:47]:
```suggestion
  otel = OpenTelemetry::Instrumentation::Grpc.client_interceptor
  SomeService::Stub.new(host, credentials, *args, **kwargs, interceptors: [otel])
```

**Review by kaylareopelle — COMMENTED** [review]
Added a few suggestions to help the markdown-related CI actions pass!

**Comment by michal-kazmierczak** [comment]
Hello @hibachrach and @kaylareopelle

I’d like to express my interest in contributing to this PR. Of course, @hibachrach has the precedence in responding to the PR review.

My potential contribution could involve either:

- making changes directly in the fork (for which I’d likely need write access), or
- implementing changes after this PR is merged.

Let me know how I can best assist!

**Comment by hibachrach** [comment]
Go for it! My personal lief has been quite busy so I haven't had the opportunity to contribute here.

**Comment by kaylareopelle** [comment]
@michal-kazmierczak, thanks for expressing your interest! And thank you, @hibachrach, for chiming in!

@michal-kazmierczak - to make things easier for you to contribute, I think we should merge this PR, knowing that the gem isn't ready for OpenTelemetry to release just yet. I'll open up an issue tomorrow with more detailed steps on what we need to take care of before we can release a new version under the open-telemetry org.

If any maintainers/approvers disagree with this approach, please let me know!

File: instrumentation/grpc/spec/example/Gemfile
Comment (kaylareopelle) [reply instrumentation/grpc/spec/example/Gemfile:11]: I'll make an issue to add a gRPC-specific example here instead. :)

**Comment by arielvalentin** [comment]
I'm ok with that as long as the unfinished gem isn't releasable or a dependency of the `all` gem.

File: instrumentation/grpc/README.md
Comment (kaylareopelle) [inline instrumentation/grpc/README.md:12]:
```suggestion
> [!CAUTION]
> This gem is not ready for release by the open-telemetry org yet, but other versions are available on RubyGems

[![Gem Version](https://badge.fury.io/rb/opentelemetry-instrumentation-grpc.svg)](https://badge.fury.io/rb/opentelemetry-instrumentation-grpc)
[![GitHub Actions CI Status](https://github.com/hibachrach/opentelemetry-instrumentation-grpc/actions/workflows/main.yml/badge.svg)](https://github.com/hibachrach/opentelemetry-instrumentation-grpc/actions?query=branch%3Amain)

OpenTelemetry instrumentation for users of the `grpc` gem

> [!WARNING]
> Right now, the gem only instruments outbound requests to gRPC services
```

**Review by kaylareopelle — APPROVED** [review]
We keep hitting Markdown Link Check failures for different links that are actually up and reporting just fine (ex: https://opensource.org/licenses/MIT). It seems like there's a timeout issue with the action itself. Instead of trying to run the failed job again, I'm going to consider this good enough and merge in the PR.

**Pull request merged**
